tests/gen/gen_auto.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gen
+
+import (
+	"math/rand"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+)
+
+const (
+	// DefaultNumDocs is the default number of documents to generate for a collection.
+	DefaultNumDocs = 10
+	// DefaultNumChildrenPerDoc is the default number of children to generate for a document.
+	DefaultNumChildrenPerDoc = 2
+
+	// DefaultStrLen is the default length of a string to generate.
+	DefaultStrLen = 10
+	// DefaultIntMin is the default minimum value of an integer to generate.
+	DefaultIntMin = 0
+	// DefaultIntMax is the default maximum value of an integer to generate.
+	DefaultIntMax = 10000
+)
+
+type docRec struct {
+	doc    map[string]any
+	docKey string
+}
+
+// AutoGenerateFromSDL generates random documents from a schema.
+func AutoGenerateFromSDL(gqlSDL string, options ...Option) ([]GeneratedDoc, error) {
+	genConfigs, err := parseConfig(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := randomDocGenerator{types: typeDefs, config: genConfigs}
+	return generator.GenerateDocs(options...)
+}
+
+// AutoGenerate generates random documents from collection definitions.
+func AutoGenerate(definitions []client.CollectionDefinition, options ...Option) ([]GeneratedDoc, error) {
+	err := validateDefinitions(definitions)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, def := range definitions {
+		typeDefs[def.Description.Name] = def
+	}
+	generator := randomDocGenerator{types: typeDefs}
+	return generator.GenerateDocs(options...)
+}
+
+type randomDocGenerator struct {
+	types        map[string]client.CollectionDefinition
+	config       configsMap
+	resultDocs   []GeneratedDoc
+	usageCounter typeUsageCounters
+	cols         map[string][]docRec
```

Comment:
todo: Related to my uncertainty RE what `docRec` is, I can't fully tell what this property is either.  If you don't feel like it is clear having renamed `docRec`, I would suggest renaming and/or documenting this property too.